design/api.md

```diff
@@ -0,0 +1,219 @@
+  {
+    "iss": "did:key:<account>",
+    "aud": "did:<server DID>",
+    "cap": { "fission:did:key:<account>": { "*": [{}] } }
+  }
+  ```
+- ```json
+  {
+    "iss": "did:<server DID>",
+    "aud": "did:key:<device>",
+    "cap": { "fission:did:key:<account>": { "*": [{}] } }
+  }
+  ```
+
+---
+
+This request registers a full account given an email verification code and delegates the account's access to given DID. In the future more UCANs may be involved in the delgation chain.
+
+
+### GET `/api/v0/account/:did`
+
+---
+
+**Authentication**:
+
+UCAN giving access to the capability `"fission:did:key:<account>": { "account/info": [{}] }` originating in `did:key:<account>`.
+The UCAN's audience needs to be `did:key:<account>`.
+
+**Response**:
+
+```ts
+{
+  email: string,
+  did: string,
+  username: string,
+}
+```
+
+---
+
+This fetches the latest account information.
+
+### GET `/api/v0/account/:username/did`
+
+---
+
+**Authentication**: *Unauthenticated*
+
+**Response**: `{ did: string }`
+
+---
+
+Use this to find out an account's DID. This information may be substituted by DoH & a DNS record in the future.
+
+
+
+
+### PUT `/api/v0/account/:did/username/:username`
+
+---
+
+**Authentication**:
+
+UCAN giving access to the capability `"fission:did:key:<account>": { "account/manage": [{}] }` originating in `did:key:<account>`.
+The UCAN's audience needs to be `did:key:<account>`.
+
+**Response**:
+
+Status 200 OK and `{ success: true }`, if successful.
+
+Status 409 Conflict and `{ success: false }`, if the username is already taken.
+
+---
+
+Change the account's username.
+
+
+### POST `/api/v0/fetch-capabilities/:did`
+
+---
+
+**Authentication**:
+
+```json
+{
+  "iss": "did:<from the path>",
+  "aud": "did:<server DID>",
+  "cap": {}, // Note: Should we introduce a capability? "did:key:zABC": { "capability/fetch": [{}] }?
+  "nnc": "<randomly generated>"
+}
+```
+
+**Request**: `{ ucans: string[] }`
+
+**Response**: `{ ucans: string[] }`
+
+---
+
+This returns a set of UCANs, that, combined with the UCAN from the request body will give the given DID access to new capabilities.
+
+In some cases this can be useful for logging in a new devices. E.g. when passkeys are available and are synced between devices, e.g. it's easy to ask a synced passkey to sign a UCAN giving the `"ucan:*": { "*": [{}] }` capability. However, this UCAN on its own doesn't give the account access to anything yet. It can use this endpoint to fetch the `fission:did:key:<account>` capability UCANs that delegate to the passkey.
+
+In the future this endpoint will be useful for fetching new capabilities from other services that were delegated to a user's account.
+
+
+## UCAN Capabilities
+
+### Resources
+
+The fission server understands one main resource, which is scoped to the `fission` URL scheme.
+
+Examples:
+- `fission:did:key:z6Mkw5d3acoQqn97UjRcGit7J7uWunixxxDKTgkr58CFHLfo`
+- `fission:did:key:z6Mkh5xiFmvPkWeHYZYnBLS4jVjsCjtpBS7D4thq1stdyEBK`
+- `fission:*`
```

Comment:
Why prefix with `fission:`? I would assume that the DID is enough to uniquely designate a resource.

Reply:
Mostly it's nice to have the resource namespaced. This allows me to do stuff like delegate `fission:*`. I use that e.g. when creating an account. I.e. on `POST /api/v0/account`, you can actually specify what DID you want the UCAN that gets returned has as audience. This allows you to create an account for a passkey without having to issue something from the passkey again. You'd then provide a UCAN proof that your passkey wants you to be able to do that.
But because before account creation the account DID doesn't exist yet, I have to use `fission:*`.

---

This... is complicated. There's quite a long decision chain which led me to this.
The root cause of this is that in the existing fission server, email verification codes are somewhat two-factor: You can only make use of an email verification from the device that you created it on. This is enforced via UCAN auth rather than the endpoints being unauthenticated.

So essentially here's what you'd do when creating an account:
`POST /api/v0/auth/email/verify` with the resource `did:key:<device DID>` and ability `email/verify`, then
`POST /api/v0/account` with the resource `fission:*` and ability `account/create`.

And these capabilities may be both delegated.
Perhaps it's clearer if I update these capabilities to use this format:
`POST /api/v0/auth/email/verify` with the resource `did:key:<device DID>` and ability `verification/create`, then
`POST /api/v0/account` with the resource `did:key:<device DID>` and ability `verification/consume`.